### PR TITLE
replace_document: check submitted ID against generated

### PR DESCRIPTION
**Pull request by spl**
* Avoids a problem with the random key type where the submitted document does not contain an `@id`.
* Consolidate `document_does_not_exist` test into `document_not_found`
* Test for `document_not_found`
* Test for WOQL `UpdateDocument` with various configurations of IDs
* Refactor `json_idgen`: Lift out `key_descriptor`, pattern match on it in `json_idgen`

### CODE REVIEW

File: src/core/api/api_error.pl [hunk 1]
```diff
@@ -166,6 +166,35 @@
...
+api_global_error_jsonld(error(document_not_found(Id), _), Type, JSON) :-
+    error_type(Type, Type_Displayed),
+    format(string(Msg), "Document not found: ~q", [Id]),
+    JSON = _{'@type' : Type_Displayed,
+             'api:status' : "api:not_found",
+             'api:error' : _{ '@type' : 'api:DocumentNotFound',
+                              'api:document_id' : Id },
+             'api:message' : Msg
+            }.
+api_global_error_jsonld(error(document_not_found(Id, Document), _), Type, JSON) :-
```
Comment (matko) [inline src/core/api/api_error.pl:178]: Do we actually generate this in a way that it appears at this level of error transform? document api endpoints should be transforming this to include the document.

Comment (matko) [reply src/core/api/api_error.pl:178]: This is what `call_catch_document_mutation` does, it checks for `known_document_error` and if the case is included, it makes sure to insert the document it was currently inserting in there. Is that appropriate in this case?

Comment (spl) [reply src/core/api/api_error.pl:178]: Yeah, this confused me for a bit. So, `document_not_found` does not actually appear in `known_document_error`.

**Review by matko — APPROVED** [review]

File: src/core/document/json.pl [hunk 2]
```diff
@@ -368,39 +368,54 @@
...
+    get_field_values(JSON, DB, Context, Fields, Values),
+    path_component([type(Base)|Path], Context, [Path_Base]),
+    idgen_hash(Path_Base, Values, Id).
+json_idgen(value_hash(Base), JSON, _DB, _Context, _Path, Id) :-
+    !,
+    get_all_path_values(JSON, Path_Values),
+    idgen_path_values_hash(Base, Path_Values, Id).
+json_idgen(Descriptor, JSON, _DB, Context, Path, Id) :-
+    (   Descriptor = random(Base)
+    ;   Descriptor = base(Base)
```
Comment (GavinMendelGleason) [inline src/core/document/json.pl:387]: This is better than before! However, stylistically, it's slightly better to have no explicit cuts, and to instead use first argument indexing, which means no fall-through case (two clauses for `random` and `base`).

Comment (spl) [reply src/core/document/json.pl]: Is it better now?

**Review by GavinMendelGleason — APPROVED** [review]

File: src/core/document/json.pl (hunk 2 above)
Comment (GavinMendelGleason) [reply src/core/document/json.pl]: Yup!

**Pull request merged**